src/huggingface_hub/serialization/_base.py

```diff
@@ -0,0 +1,168 @@
+# Copyright 2024 The HuggingFace Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Contains helpers to split tensors into shards."""
+from dataclasses import dataclass, field
+from typing import Any, Callable, Dict, List, Optional, TypeVar
+
+from .. import logging
+
+
+TensorT = TypeVar("TensorT")
+TensorSizeFn_T = Callable[[TensorT], int]
+StorageIDFn_T = Callable[[TensorT], Optional[Any]]
+
+MAX_SHARD_SIZE = 5_000_000
+FILENAME_PATTERN = "model{suffix}.safetensors"
+
+logger = logging.get_logger(__file__)
+
+
+@dataclass
+class StateDictSplit:
+    is_sharded: bool = field(init=False)
+    metadata: Dict[str, Any]
+    filename_to_tensors: Dict[str, List[str]]
+    tensor_to_filename: Dict[str, str]
+
+    def __post_init__(self):
+        self.is_sharded = len(self.filename_to_tensors) > 1
+
+
+def split_state_dict_into_shards(
+    state_dict: Dict[str, TensorT],
+    *,
+    get_tensor_size: TensorSizeFn_T,
+    get_storage_id: StorageIDFn_T = lambda tensor: None,
+    filename_pattern: str = FILENAME_PATTERN,
+    max_shard_size: int = MAX_SHARD_SIZE,
+) -> StateDictSplit:
```

Comment:
If I'm not mistaken this is more a factory function rather than an end-user function; this is to be implemented by third-parties exactly like you did with torch/tf/numpy

Its current name made me believe that it was a wrapper over all other methods rather than the backend of all of these methods; would renaming it to make this clearer be an option? Something like `split_state_dict_into_shards_meta` or `split_state_dict_into_shards_abstraction` (not the best name)?

Reply:
Additionally, maybe such a wrapper function that would automatically send to the detected backend could be nice. More of a nice-to-have than something groundbreaking though.

Reply:
What about `split_state_dict_into_shards_factory`? 

> maybe such a wrapper function that would automatically send to the detected backend could be nice

That was in the initial version of this PR but in the end I don't think it's worth it. Third-party libraries integrating it should be aware of which framework they are using IMO. If we have demand for it we can reevaluate though.

The "problem" with auto-detection is that we might need some magic to detect the framework type without actually importing unnecessary libs. Example: I have torch and tensorflow installed. I want to test from which framework is my tensor. A "naive" logic like the following one might unnecessarily import the whole torch or tensorflow library just to make a single test.

```py
if is_torch_available():
    import torch
    if isinstance(tensor, torch.Tensor):
        return ... # torch specific
if is_tf_available():
    import tensorflow as tf
    if isinstance(tensor, tf.Tensor):
        return ... # tf specific
```

I haven't thought too much of a workaround for that but keeping methods separated seems ok to me anyway.